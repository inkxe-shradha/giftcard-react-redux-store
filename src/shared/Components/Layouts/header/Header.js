import React from "react";
import { AppBar, Grow, MenuItem, MenuList, Paper, Popper } from "@mui/material";
import Toolbar from "@mui/material/Toolbar";
import Typography from "@mui/material/Typography";
import { ClickAwayListener } from "@mui/material";
import Button from "@mui/material/Button";
import IconButton from "@mui/material/IconButton";
import MenuIcon from "@mui/icons-material/Menu";
import { Link, useLocation, useNavigate } from "react-router-dom";
import { makeStyles } from "@mui/styles";
import AccountCircleIcon from "@mui/icons-material/AccountCircle";
import Navbar from "../../../../components/Navbar/Navbar";
import clsx from "clsx";
import menuItems from "../../../../config/menuList";
import PropTypes from "prop-types";
const drawerWidth = 240;

const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1,
    display: "flex",
  },
  menuButton: {
    marginRight: theme.spacing(2),
  },
  title: {
    flexGrow: 1,
    cursor: "pointer",
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    transition: theme.transitions.create(["width", "margin"], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    [theme.breakpoints.up("md")]: {
      width: "100% !important",
    },
  },
  hide: {
    display: "none",
  },
  drawer: {
    width: drawerWidth,
    flexShrink: 0,
    whiteSpace: "nowrap",
  },
  drawerPaper: {
    width: drawerWidth,
  },
  navDrawer: {
    width: drawerWidth,
    // Page size nave drawer options
    [theme.breakpoints.down("md")]: {
      width: drawerWidth,
      flexShrink: 0,
      display: "none",
    },
    [theme.breakpoints.up("md")]: {
      width: drawerWidth,
      flexShrink: 0,
    },
  },
  drawerOpen: {
    width: drawerWidth,
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  drawerClose: {
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    overflowX: "hidden",
    width: theme.spacing(7) + 1,
    [theme.breakpoints.up("sm")]: {
      width: theme.spacing(9) + 1,
    },
  },
  appBarShift: {
    marginLeft: drawerWidth,
    width: `calc(100% - ${drawerWidth}px)`,
    transition: theme.transitions.create(["width", "margin"], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  toolbar: {
    display: "flex",
    alignItems: "center",
    justifyContent: "flex-end",
    padding: theme.spacing(0, 1),
    // necessary for content to be below app bar
    ...theme.mixins.toolbar,
  },
  content: {
    flexGrow: 1,
    padding: theme.spacing(3),
  },
  signedInUser: {
    padding: theme.spacing(1),
    cursor: "pointer",
  },
  hiddenSmallScreen: {
    [theme.breakpoints.down("md")]: {
      display: "none",
    },
  },

  hiddenBigScreen: {
    [theme.breakpoints.up("md")]: {
      display: "none !important",
    },
  },
  hideMenu: {
    display: "none !important",
  },
}));

const Header = ({ user, authStatus, logOut }) => {
  const classes = useStyles();
  const location = useLocation();
  const [open, setOpen] = React.useState(false);
  const navigate = useNavigate();

  const anchorRef = React.useRef(null);
  const [menuOpen, setMenuOpen] = React.useState(false);
  React.useEffect(() => {
    return () => {
      anchorRef && anchorRef.current && anchorRef.current.focus();
    };
  }, []);
  const handelToggle = () => {
    setMenuOpen(!menuOpen);
  };
  const handleListKeyDown = (event) => {
    if (event.key === "Tab") {
      event.preventDefault();
      setOpen(false);
    }
  };
  const handleClose = (event) => {
    if (anchorRef.current && anchorRef.current.contains(event.target)) {
      return;
    }
    setMenuOpen(false);
  };
  //Profile & Logout Menu
  const handleToggleBtn = () => {
    setOpen((prevOpen) => !prevOpen);
  };

  const handelLogOut = (event) => {
    logOut();
    handleClose(event);
  };

  const handelProfileClick = (event) => {
    setMenuOpen(false);
    navigate("/profile");
  };
  return (
    <div className={classes.root}>
      {/* AppBar Implementation */}
      <AppBar
        className={clsx(classes.appBar, {
          [classes.appBarShift]: open,
        })}
        position="fixed"
      >
        <Toolbar>
          <IconButton
            onClick={handleToggleBtn}
            edge="start"
            className={clsx(classes.menuButton, classes.hiddenBigScreen)}
            color="inherit"
            aria-label="menu"
          >
            <MenuIcon />
          </IconButton>
          <Typography
            variant="h6"
            onClick={() => navigate("/")}
            className={classes.title}
          >
            YoYo Gifting
          </Typography>
          <div
            className={clsx(classes.hiddenSmallScreen, {
              ["d-none"]: !authStatus,
            })}
          >
            {menuItems.map((ele) => {
              return (
                <Button
                  key={ele.id}
                  color={location.pathname === ele.path ? "primary" : "inherit"}
                  component={Link}
                  to={ele.path}
                >
                  {" "}
                  {ele.name}{" "}
                </Button>
              );
            })}
          </div>
          <div className={classes.signedInUser}>
            <Typography color="secondary">
              Hello, {user?.name || "Guest!"}
            </Typography>
          </div>

          <div
            className={clsx(classes.signedInUser, {
              [classes.hideMenu]: !authStatus,
            })}
          >
            <AccountCircleIcon
              ref={anchorRef}
              aria-controls={menuOpen ? "menu-list-grow" : undefined}
              aria-haspopup="true"
              onClick={handelToggle}
            />
            <Popper
              open={menuOpen}
              anchorEl={anchorRef.current}
              role={"Menu list grow"}
              transition
              disablePortal
            >
              {({ TransitionProps, placement }) => (
                <Grow
                  {...TransitionProps}
                  style={{
                    transformOrigin:
                      placement === "bottom" ? "center top" : "center bottom",
                  }}
                >
                  <Paper>
                    <ClickAwayListener onClickAway={handleClose}>
                      <MenuList
                        autoFocusItem={menuOpen}
                        id="menu-list-grow"
                        onKeyDown={handleListKeyDown}
                      >
                        <MenuItem onClick={handelProfileClick}>
                          Profile
                        </MenuItem>
                        <MenuItem onClick={handelLogOut}>Logout</MenuItem>
                      </MenuList>
                    </ClickAwayListener>
                  </Paper>
                </Grow>
              )}
            </Popper>
          </div>
          {!authStatus && (
            <div>
              <Button LinkComponent={Link} to="/login" color="inherit">
                Login
              </Button>
            </div>
          )}
        </Toolbar>
      </AppBar>
      <Navbar {...classes} open={open} handleDrawerClose={handleToggleBtn} />
    </div>
  );
};

// props validation
Header.propTypes = {
  user: PropTypes.object.isRequired,
  authStatus: PropTypes.bool.isRequired,
  logOut: PropTypes.func.isRequired,
};

export default Header;
