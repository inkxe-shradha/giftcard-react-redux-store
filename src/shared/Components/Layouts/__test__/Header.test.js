import { mount } from "enzyme";
import { ThemeProvider } from "@mui/system";
import { BrowserRouter as Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { createTheme } from "@mui/material";
import Header from "../header/Header";
import { render, screen } from "@testing-library/react";

const theme = createTheme();
describe("Header Test Suite Testing", () => {
  const customRender = (component, location) => {
    const history = createMemoryHistory();
    history.push(location);
    return render(
      <ThemeProvider theme={theme}>
        <Router location={history.location} navigator={history}>
          {component}
        </Router>
      </ThemeProvider>
    );
  };
  const mountRender = (component, location) => {
    const history = createMemoryHistory();
    history.push(location);
    return mount(
      <ThemeProvider theme={theme}>
        <Router location={history.location} navigator={history}>
          {component}
        </Router>
      </ThemeProvider>
    );
  };

  const layoutState = {
    user: {
      name: "Shradha suman praharaj",
      email: "",
      phone: "",
    },
    authStatus: false,
    logOut: jest.fn(),
  };

  it("Should create a snapshot of the header if the user is not authenticated", () => {
    const { asFragment } = customRender(
      <Header
        user={layoutState.user}
        authStatus={false}
        logOut={layoutState.logOut}
      />,
      "/"
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should be show login if the user is not authenticated", () => {
    customRender(
      <Header
        user={layoutState.user}
        authStatus={layoutState.authStatus}
        logOut={layoutState.logOut}
      />,
      "/"
    );
    expect(screen.getByText(/login/i)).toBeTruthy();
  });

  it("Should be show Gift received and Gift Sent button if the user is authenticated", () => {
    const view = mountRender(
      <Header
        user={layoutState.user}
        authStatus={true}
        logOut={layoutState.logOut}
      />,
      "/"
    );
    expect(view.text().includes("Gift Received")).toBeTruthy();
    expect(view.text().includes("Gift Sent")).toBeTruthy();
    expect(view.text().includes("Home")).toBeTruthy();
  });

  it("Should be show logout button if the user is authenticated", () => {
    const view = mountRender(
      <Header
        user={layoutState.user}
        authStatus={true}
        logOut={layoutState.logOut}
      />,
      "/"
    );
    // click button
    view.find("svg[data-testid='AccountCircleIcon']").simulate("click");
    expect(view.text().includes("Logout")).toBeTruthy();
  });
});
