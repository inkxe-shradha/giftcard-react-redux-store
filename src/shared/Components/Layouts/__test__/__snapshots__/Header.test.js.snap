// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Test Suite Testing Should create a snapshot of the header if the user is not authenticated 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed makeStyles-appBar-4 mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
      >
        <button
          aria-label="menu"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium makeStyles-menuButton-2 makeStyles-hiddenBigScreen-16 css-1d6w9lk-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root MuiTypography-h6 makeStyles-title-3 css-2ulfj5-MuiTypography-root"
        >
          YoYo Gifting
        </h6>
        <div
          class="makeStyles-hiddenSmallScreen-15 d-none"
        >
          <a
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            href="/"
            tabindex="0"
          >
             Home 
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <a
            class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-1y942vo-MuiButtonBase-root-MuiButton-root"
            href="/gift-received"
            tabindex="0"
          >
             Gift Received 
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <a
            class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-1y942vo-MuiButtonBase-root-MuiButton-root"
            href="/gift-sent"
            tabindex="0"
          >
             Gift Sent 
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
        <div
          class="makeStyles-signedInUser-14"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-18brw8f-MuiTypography-root"
          >
            Hello, Shradha suman praharaj
          </p>
        </div>
        <div
          class="makeStyles-signedInUser-14 makeStyles-hideMenu-17"
        >
          <svg
            aria-haspopup="true"
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AccountCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </div>
        <div>
          <a
            class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-1y942vo-MuiButtonBase-root-MuiButton-root"
            href="/login"
            tabindex="0"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </header>
    <nav
      class="makeStyles-navDrawer-8"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 makeStyles-drawerPaper-7 css-joixkg-MuiPaper-root"
      />
    </nav>
  </div>
</DocumentFragment>
`;
